### PR TITLE
interpret: make read/write methods generic

### PR DESCRIPTION
**Pull request by RalfJung**
Instead of always having to call `into()` to convert things to `PlaceTy`/`OpTy`, make the relevant methods generic. This also means that when we read from an `MPlaceTy`, we avoid creating an intermediate `PlaceTy`.

This makes it feasible to remove the `Copy` from `MPlaceTy`. All the other `*Ty` interpreter types already had their `Copy` removed a while ago so this is only consistent. (And in fact we had one function that accidentally took `MPlaceTy` instead of `&MPlaceTy`.)

### CODE REVIEW

File: compiler/rustc_const_eval/src/const_eval/valtrees.rs [hunk 1]
```diff
@@ -290,7 +290,7 @@ pub fn valtree_to_const_value<'tcx>(
             debug!(?place);
 
             valtree_into_mplace(&mut ecx, &mut place, valtree);
-            dump_place(&ecx, place.into());
+            dump_place(&ecx, place.clone().into());
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/valtrees.rs:293]: This seems rather silly: we have a whole bunch of `clone().into()` just for all this dumping. Can't we just inline the `trace!("{:?}", ecx.dump_place(*place))` everywhere? Then at least the clone/into is avoided when `trace!` is disabled...

Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/valtrees.rs:293]: actually all the callers to `dump_place` pass a `MplaceTy` and `.into()` it. So we may as well do that `.into()` within the function 😆

**Comment by RalfJung** [comment]
Found some more unnecessary `&mut` around `valtree_into_mplace`

@bors r=oli-obk

**Pull request merged**